src/state_db.rs

```diff
@@ -454,9 +451,293 @@ impl StateDB {
+                    let (cached_value, deleted, exists) = writer.get(&key);
+                    if exists && !deleted {
+                        let buffer = JsBuffer::external(&mut ctx, cached_value);
+                        vec![ctx.null().upcast(), buffer.upcast()]
+                    } else if deleted {
+                        vec![ctx.error("No data")?.upcast()]
+                    } else if let Ok(value) = &value {
+                        // if found the value of the key then insert into cache
+                        if value.is_some() {
+                            let temp_value = value.as_ref().unwrap().to_vec();
+                            let pair = SharedKVPair::new(&key, &temp_value);
+                            writer.cache_existing(&pair);
+                            let buffer = JsBuffer::external(&mut ctx, temp_value);
+                            vec![ctx.null().upcast(), buffer.upcast()]
+                        } else {
+                            vec![ctx.error("No data")?.upcast()]
+                        }
+                    } else {
+                        let err = value.err().unwrap();
+                        vec![ctx.error(&err)?.upcast()]
+                    }
+                };
+
+                let this = ctx.undefined();
+                let callback = cb.into_inner(&mut ctx);
+                callback.call(&mut ctx, this, args)?;
+                Ok(())
+            });
+        })
+    }
+
+    fn cache_to_js_array<'a, C: Context<'a>>(
+        ctx: &mut C,
+        values: &Cache,
+    ) -> NeonResult<Handle<'a, JsArray>> {
+        let res_values = ctx.empty_array();
+        for (i, kv) in values.iter().enumerate() {
+            let temp_pair = KVPair::new(kv.0, kv.1);
+            let object = Self::pair_to_js_object(ctx, &temp_pair)?;
+            res_values.set(ctx, i as u32, object)?;
+        }
+
+        Ok(res_values)
+    }
+
+    fn pair_to_js_object<'a, C: Context<'a>>(
+        ctx: &mut C,
+        pair: &KVPair,
+    ) -> NeonResult<Handle<'a, JsObject>> {
+        let obj = ctx.empty_object();
+        let key = JsBuffer::external(ctx, pair.key_as_vec());
+        obj.set(ctx, "key", key)?;
+        let value = JsBuffer::external(ctx, pair.value_as_vec());
+        obj.set(ctx, "value", value)?;
+
+        Ok(obj)
+    }
+
+    fn range(
+        &self,
+        cb: Root<JsFunction>,
+        writer: ArcMutex<state_writer::StateWriter>,
+        options: options::IterationOption,
+    ) -> Result<(), mpsc::SendError<options::DbMessage>> {
+        self.common.send(move |conn, channel| {
+            let values = conn
+                .iterator(utils::get_iteration_mode(&options, &mut vec![], true))
+                .map(|(k, v)| KVPair::new(&k, &v))
+                .collect::<Vec<KVPair>>();
+            channel.send(move |mut ctx| {
+                let result = {
+                    let mut writer = writer.lock().unwrap();
```

Comment:
You may release writer in similar way which is done in `delete_key` function.

Reply:
same as `args`